Scijava/scijava javadoc parser/javadoc to yaml

gselzer opened the pull request:
This PR introduces a new SciJava annotation processor to replace the usage of therapi. The processor creates YAML files describing all implementations (identified with an `@implNote` tag), for use in SciJava 3.

The YAML created by the processor is designed for usage beyond Ops, and for robust declaration of implementations from arbitrary, cross-langauge sources. Each implementation must contain the following:
* An implementation type. Implementations are binned by their implementation type into `type.yaml` files (for example, `op.yaml` contains all of the `op` implementations discovered by the processor).
* An implementation source, a `URI` that some yaml reader could use to get an instance of the implementation. For example, this could be a FQCN in Java, along with a scheme identifying that the URI is a Java class scheme.
* A map of tags, containing *type-specific* information. For ops, this includes the op's name, it's parameter information, priority, etc.

As an example, here's an Op:
```java
	/**
	 * A Field Op
	 * @implNote op name=example.mul
	 * @input a the first double
	 * @input b the first double
	 * @output the product
	 */
	public final BiFunction<Double, Double, Double> multiply = (a, b) -> a * b;
```
And here is the generated YAML, placed into `op.yaml`:
```yaml
  - op:
      name: example.mul
      priority: 0.0
      source: javaField:/org.scijava.ops.engine.yaml.ops.YAMLFieldOp$multiply/v0.0
      parameters:
        - input: in1
          description: The first input
        - input: in2
          description: The second input
        - output: out
          description: The product
```
Although the build completes and passes tests, this PR is a draft as there are a couple things left to do:
* Polish the annotation processor, and the yaml readers, and the schema we're currently using - while ImageJ Ops2 tests it, I'm sure there are still edge cases and inefficiencies.
* Determine whether we need @hinerm's ruby script in `pom.xml` to merge the `*.yaml` files between source and tests - @ctrueden seemed to think that we don't need to do that. My experience is that it is necessary, though..

gselzer committed:
Parse docs with our own annotation processor

ctrueden commented:
What I suggested was to make the path to the YAML for `src/test` distinct, like `op.test.yaml`, and check whether the `getResources` call is able to discover it then. I.e. is the problem a clash between multiple YAML files of the same path in the same project? Or is the issue merely discovery of resources in `target/test-classes` in general, maybe due to JPMS issues? If the former, using a different name works around the problem.

gselzer commented:
Closes scijava/scijava#65

ctrueden reviewed (changes requested):
Thank you so much for working on this, @gselzer! I'm excited to have ops driven uniformly by YAML metadata. But it looks like this PR is not quite ready for review yet:

* The description has empty code blocks for the op and corresponding YAML.
* The commit history has WIPs.

gselzer commented:
@ctrueden I edited my PR to add the example Op and resulting YAML, and squashed together all of the WIP commits into a single commit, because things really weren't passing the build until the very end. Should be good to go now!

gselzer committed:
Name javadoc'd Op params after name in method

hinerm commented:
Build with:
`mvn clean install -pl scijava/scijava-javadoc-parser -am; mvn clean install`

Questions:
- Can everything build in on multi-module build? Or does the javadoc parser need to be in its own component?
- Is there an easier way to debug the annotation processor than logging output?

hinerm commented:
@gselzer what are the benefits to replacing the therapi mechanism for harvesting Javadoc ops? I didn't see a corresponding issue for this PR.

gselzer commented:
> @gselzer what are the benefits to replacing the therapi mechanism for harvesting Javadoc ops? I didn't see a corresponding issue for this PR.

* Some benefits are described in scijava/scijava#65.
* A further benefit is that we no longer have to scan the entire classpath. Right now, there is no way to ask therapi "give me all javadoc'd elements that contain an `@implNote` annotation - we have to scan the *entire* classpath to do that, which actually breaks down when the classpath isn't fully populated e.g. in Fiji. It's also expensive.

There may be more, which I'm forgetting...

hinerm commented:
@gselzer I've rebased this branch against `main`. Currently all the `imagej-ops` tests are failing, despite running the separate ` mvn clean install -pl scijava/scijava-javadoc-parser -am; mvn clean install` commands. Can you take a look?

Also are the op parameter names being retained? Can you point me to an example of where they are used?

gselzer commented:
@hinerm looking now, it seems to be an issue with type variables. There are some identifiers in the imagej ops2 yaml that have type variables, such as:
```yaml
net.imagej.ops2.copy.Copiers.<T>copyType(T, T)
```
unfortunately, we are not correctly reifying the parameter types. I think that the solution is just to erase the type variables to their bounds. I'll let you know when I have a fix.

gselzer committed:
Fix YAML issues

hinerm commented:
@gselzer based on discussion today:
- [ ] Identify where exceptions are being eaten for invalid op javadoc (methods without a type) and see if we can get that to fail compilation

gselzer committed:
Make minor fixes

gselzer commented:
> @gselzer based on discussion today:
> 
>     * [ ]  Identify where exceptions are being eaten for invalid op javadoc (methods without a type) and see if we can get that to fail compilation

So, looking back at this, I think we should avoid failing compilation - the javadoc parser is designed to be more general than Ops, and thus we may not always want to conform to the Javadoc requirements for an Op.

What we *can* do is ensure that the tests fail *because* an Op written as a method doesn't have a type hint. Right now, if we miss a type hint, the errors look like this:
```
[ERROR] Tests run: 4, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 0.173 s <<< FAILURE! - in org.scijava.ops.engine.yaml.YAMLOpTest
[ERROR] testYAMLFunction  Time elapsed: 0.075 s  <<< ERROR!
java.lang.IllegalArgumentException: java.lang.RuntimeException: java.lang.RuntimeException: Op org.scijava.ops.engine.yaml.ops.YAMLMethodOp.subtract(java.lang.Double,java.lang.Double) could not be loaded: Could not load class null
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.yaml.YAMLOpTest.testYAMLFunction(YAMLOpTest.java:69)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Op org.scijava.ops.engine.yaml.ops.YAMLMethodOp.subtract(java.lang.Double,java.lang.Double) could not be loaded: Could not load class null
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.yaml.YAMLOpTest.testYAMLFunction(YAMLOpTest.java:69)
Caused by: java.lang.RuntimeException: Op org.scijava.ops.engine.yaml.ops.YAMLMethodOp.subtract(java.lang.Double,java.lang.Double) could not be loaded: Could not load class null
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.yaml.YAMLOpTest.testYAMLFunction(YAMLOpTest.java:69)
Caused by: java.lang.NullPointerException: Cannot invoke "String.equals(Object)" because "name" is null
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.yaml.YAMLOpTest.testYAMLFunction(YAMLOpTest.java:69)

[ERROR] testYAMLClass  Time elapsed: 0.001 s  <<< ERROR!
org.scijava.ops.api.features.OpMatchingException: No MatchingRoutine was able to produce a match!
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.yaml.YAMLOpTest.testYAMLClass(YAMLOpTest.java:48)

[ERROR] testYAMLInnerClass  Time elapsed: 0.001 s  <<< ERROR!
org.scijava.ops.api.features.OpMatchingException: No MatchingRoutine was able to produce a match!
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.yaml.YAMLOpTest.testYAMLInnerClass(YAMLOpTest.java:55)

[ERROR] testYAMLMethod  Time elapsed: 0 s  <<< ERROR!
org.scijava.ops.api.features.OpMatchingException: No MatchingRoutine was able to produce a match!
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.yaml.YAMLOpTest.testYAMLMethod(YAMLOpTest.java:62)

```

I think this is good enough, for now.

gselzer commented on .github/build.sh:
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh
 curl -fsLO https://raw.githubusercontent.com/scijava/scijava-scripts/master/ci-build.sh
-mvn -Djavadoc.skip -pl scijava/scijava-taglets clean install
+mvn -Djavadoc.skip -pl scijava/scijava-taglets,scijava/scijava-javadoc-parser -am clean install
```
I'd really prefer to get these modules out of the incubator ASAP - @ctrueden the taglets, at least, seem pretty polished - should we consider migrating them?

gselzer commented on imagej/imagej-ops2/src/main/java/net/imagej/ops2/copy/Copiers.java:
```diff
@@ -113,27 +113,27 @@ public static <T extends NativeType<T>, A extends ArrayDataAccess<A>> void copyA
+//            throw new IllegalArgumentException("input and output must be of the same dimensions!");
+//        Computers.Arity1<Iterable<T>, Iterable<T>> mapped = Maps.ComputerMaps.Iterables.liftBoth(copier);
+//        mapped.compute(input, output);
+//    }
```
Let's either delete this or uncomment it  - either we need it, or we don't!

gselzer replied:
Fixed in 0520fede

gselzer commented on pom.xml:
```diff
@@ -111,20 +111,20 @@
 					<fork>true</fork>
+					<showWarnings>true</showWarnings>
 					<compilerArgs>
 						<arg>-Ajavadoc.packages="${therapi.packages}"</arg>
```
Should rename the property to something else, although I can't think of any suggestions quickly.

gselzer replied:
Renamed to `parse.ops` in 296c41ac

gselzer commented on pom.xml:
```diff
@@ -201,19 +206,20 @@
-									testPluginsPath = "#{basedir}/target/test-classes/#{pluginsPath}"
-									mainPlugins = read_plugins(mainPluginsPath)
-									testPlugins = read_plugins(testPluginsPath)
+									['op.yaml', 'META-INF/json/org.scijava.plugin.Plugin'].each do |pluginsPath|
```
Can we make this work with all yaml files? The javadoc parser could theoretically work with any `implNote` types.

gselzer replied:
The following section will get all files ending in `.yaml`:

```ruby
Dir.glob("#{Dir.getwd}/.yaml")
```

If we can concat `META-INF/json/org.scijava.plugin.Plugin` to the end, we *should* be good!

gselzer replied:
Fixed with 9d032d0e

gselzer reviewed (commented):
Reviewed a little bit!

gselzer commented on scijava/scijava-common3/src/main/java/org/scijava/common3/Classes.java:
```diff
@@ -183,8 +183,14 @@ public static Class<?> load(final String name, final ClassLoader classLoader,
+			else {
+				if (quietly) return null;
+				throw iae(t, "Cannot load class: " + className);
+			}
```
This is a hack to help discover inner classes. We should probably mark it as such, at the very least, and arguably not even leave it here in the catch.

gselzer replied:
Moved the HACK out to where it  is needed (JavaMethodYAMLInfoCreator) in 2bf8bec9

gselzer commented on scijava/scijava-javadoc-parser/README.md:
We should change the readme to reflect the new project

gselzer replied:
And should also remember to attribute therapi for the starting code. Check their license

gselzer commented on scijava/scijava-javadoc-parser/pom.xml:
```diff
@@ -9,9 +9,9 @@
 
-	<name>SciJava Discovery Therapi</name>
+	<name>SciJava Javadoc Parser</name>
 	<description>SciJava Discovery Therapi: A therapi-aided Discovery implementation</description>
```
Need to change the description

gselzer replied:
Should also make sure that SciJava Discovery Therapi doesn't show up anywhere in this POM

gselzer replied:
It's no longer in the POM, I purged the old name in b8f0cc741

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplClassData.java:
```diff
@@ -0,0 +1,109 @@
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+
+public class ImplClassData implements ImplData {
```
Some Javadoc would be nice

gselzer replied:
And the class itself could use formatting

gselzer replied:
Done!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplData.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public interface ImplData {
```
This class also needs docs

gselzer replied:
Doc-ified!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplClassData.java:
```diff
@@ -0,0 +1,109 @@
+						}
+						break;
+				default:
+					if (description == null) {
```
I feel like this switch could be made much more general - the cases seem very WET.

gselzer replied:
It's much more general now!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplClassData.java:
```diff
@@ -0,0 +1,109 @@
+			srcString = srcString.substring(0, badPeriod) + '$' + srcString.substring(badPeriod+1);
+			parent = parent.getEnclosingElement();
+		}
+		this.source = "javaClass:/" + URLEncoder.encode(srcString, StandardCharsets.UTF_8);
```
Might want to refactor this into a separate method - feels like an SRP violation

gselzer replied:
It's in its own method now!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplFieldData.java:
```diff
@@ -0,0 +1,105 @@
+
+import javax.lang.model.element.Element;
+
+public class ImplFieldData implements ImplData {
```
Needs docs

gselzer replied:
Has docs!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplFieldData.java:
```diff
@@ -0,0 +1,105 @@
+						description = section;
+					}
+					break;
+			}
```
Carbon copy of what is in `ImplClassData` - should we add an abstract base class?

gselzer replied:
I did that!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplFieldData.java:
```diff
@@ -0,0 +1,105 @@
+				Collectors.toList()));
+		map.putAll(implNotes);
+		return map;
+	}
```
Why do Fields have extra tags?

gselzer replied:
I...don't think they do anymore?

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplMethodData.java:
```diff
@@ -0,0 +1,142 @@
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.util.Types;
+
+public class ImplMethodData implements ImplData{
```
Needs docs

gselzer replied:
Has docs now!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplMethodData.java:
```diff
@@ -0,0 +1,142 @@
+			switch(elements[0]) {
+				case "param":
+					// Ignore type variables
+					if (elements[1].contains("<.*>")) continue;
```
Kleene + might be more accurate

gselzer replied:
So now I have the following HACK, which works in our limited test scenarios, but I'd love if we could do better.
```java
	/**
	 * HACK to find type variable param tags
	 * For a parameter tag, returns {@code true} iff the following tag is a type variable tag.
	 * Type variable tags start with a greater than sign, and then has a string of letters, and then a less than sign.
	 * @param tag the string following an param tag
	 * @return true iff the tag is an param tag
	 */
	private boolean paramIsTypeVariable(String tag) {
		// TODO: Why doesn't Pattern.matches(".*<\\p{L}>.*", tag) work??
		if (tag.charAt(0) != '<') return false;
		for(int i = 1; i < tag.length(); i++) {
			char c = tag.charAt(i);
			if (Character.isLetter(c)) continue;
			return c == '>';
		}
		return false;
	}

```

@ctrueden @hinerm if you have any ideas here, I'd be all ears.

ctrueden replied:
```java
Pattern.matches("<\\s*[A-Za-z_]\\w*\\s*>", tag)
```
?

gselzer replied:
My hack works well enough, so I'm not going to spend a ton of time debugging

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ImplMethodData.java:
```diff
@@ -0,0 +1,142 @@
+								}
+							}
+						}
+						break;
```
This should probably be in `RuntimeJavadocHelper` as a static utility method

gselzer replied:
It's just been moved to a base class.

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+
+import org.yaml.snakeyaml.Yaml;
+
+public class JavadocAnnotationProcessor extends AbstractProcessor {
```
Needs docs

gselzer replied:
Added docs and renamed!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnvironment) {
+        processingEnv.getMessager().printMessage(Diagnostic.Kind.NOTE, "Processing!!!!!");
```
Can we come up with a more informational message? :sweat_smile:

gselzer replied:
Added a better message

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+                    }
+                }
+            }
+        }
```
This may be unnecessary

gselzer replied:
I've greatly simplified it

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+
+        // Retain Javadoc for classes that match this predicate
+        final PackageFilter packageFilter =
+                packagesOption == null ? new PackageFilter() : new PackageFilter(packagesOption);
```
Honestly, how much do we care about the option to block off some packages? This isn't like the Javadoc parser, where *everything* is retained within one file per `.java` file - we're condensing down to one file per implementation type. It's also unlike the Javadoc parser in that we're only retaining things with the `@implNote` tag - it's more opt-in than before.  I'm really tempted to remove the option to specify packages.

gselzer replied:
Removed the option to specify packages in 8d0a62db.

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+            }
+        }
+
+        return false;
```
We should define *why* we're returning `false` here with a comment

gselzer replied:
Added a comment in 5c898546

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+
+    private static boolean isRetainJavadocAnnotation(TypeElement annotation) {
+        return true;
+    }
```
Not sure what logic this contained in therapi, but it's worthless here - just remove the `if` conditional aboce

gselzer replied:
It's been removevd

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+
+    static {
+        RECORD.ifPresent(elementKindsToInspect::add);
+    }
```
`findValueOf` is only used within this selected block - why don't we just inline the logic within the `static` block?

gselzer replied:
I actually just removed the ability to scan records - we can add it back later if needed, since we aren't trying to be general in discovery anymore - only focused on ops

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+                processingEnv.getMessager()
+                    .printMessage(Diagnostic.Kind.ERROR, "Javadoc retention failed; " + ex, element);
+                throw new RuntimeException("Javadoc retention failed for " + element, ex);
+            }
```
This naming seems misleading to me - theoretically, `element` could be any of the 4 types of `ElementKind` specified above - not just a `Class` - maybe we should rename the method? Or should we narrow the list of kinds to inspect?

gselzer replied:
Method has been renamed

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+            return;
+        }
+        TypeElement classElement = (TypeElement) element;
+        yamlJavadocBuilder.getClassJavadocAsYamlOrNull(classElement) //
+            .forEach(impl -> {
+                yamlMap.computeIfAbsent(impl.type(), i -> new ArrayList<>()).add(Collections.singletonMap(impl.type(), impl.dumpData()));
+            });
+    }
```
This is *super* confusing, and prohibitive to read

gselzer replied:
It's a lot better now!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/JavadocAnnotationProcessor.java:
```diff
@@ -0,0 +1,195 @@
+        String relativeName = type + ".yaml";
+        return processingEnv.getFiler()
+            .createResource(StandardLocation.CLASS_OUTPUT, "", relativeName);
+    }
+    private static PackageElement getPackageElement(Element element) {
```
Format this file

gselzer replied:
Formatted!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/PackageFilter.java:
```diff
@@ -0,0 +1,82 @@
+import javax.lang.model.element.QualifiedNameable;
+
+class PackageFilter {
+
```
Should give this file proper attribution - we didn't write it, it came directly from therapi

gselzer replied:
Gone now!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/ParameterTagData.java:
```diff
@@ -0,0 +1,82 @@
+
+public class ParameterTagData {
+
+	public enum IO_TYPE {
+		INPUT,
+		OUTPUT,
+		MUTABLE,
+		CONTAINER
+	}
```
Is this enum really needed?

gselzer replied:
I don't really see any reason to move it!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/RuntimeJavadocHelper.java:
```diff
@@ -0,0 +1,153 @@
+import java.util.regex.Pattern;
+
+public class RuntimeJavadocHelper {
+    public static final String INIT = "<init>";
```
This is never used - I'd delete anything in this file that we aren't using

gselzer replied:
The whole file is gone now!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/RuntimeJavadocHelper.java:
```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2015 David Nault and contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
We should rewrite the LICENSE here - @ctrueden is there a protocol for doing so?

ctrueden replied:
You can model it after `Types.java`:

1. Put a commented notice about the original license after our initial license block; and
2. Make sure that commented notice tells people to check the `NOTICE`/`NOTICE.txt` for additional details.

As an aside, that `Types.java` needs to be adjusted to point to an actually existing file called either `NOTICE` or `NOTICE.txt` or `THIRDPARTY.txt`. See [scijava-common's `NOTICE.txt`](https://github.com/scijava/scijava-common/blob/scijava-common-2.96.0/NOTICE.txt).

gselzer replied:
Added with 0fa1d47d

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/RuntimeJavadocHelper.java:
```diff
@@ -0,0 +1,153 @@
+public class RuntimeJavadocHelper {
+    public static final String INIT = "<init>";
+
+    public static final Pattern
+        blockSeparator = Pattern.compile("^\\s*@(?=\\S)", Pattern.MULTILINE);
+
+    public static final Pattern tagElementSeparator = Pattern.compile("\\s*,*\\s+(?=(?:[^']*'[^']*')*[^']*$)");
```
Some javadoc describing what these patterns do would be great

gselzer replied:
They are described now!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/RuntimeJavadocHelper.java:
```diff
@@ -0,0 +1,153 @@
+        return list == null ? Collections.emptyList() : unmodifiableList(new ArrayList<>(list));
+    }
+
+    public static <T> T requireNonNull(T object) {
+        if (object == null) {
+            throw new NullPointerException();
+        }
+        return object;
+    }
```
Replace usage with `Objects.requireNonNull`

gselzer replied:
It's gone now!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/YamlJavadocBuilder.java:
```diff
@@ -0,0 +1,108 @@
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+
+class YamlJavadocBuilder {
```
Needs docs

gselzer replied:
Nope - it's no longer a file!

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/YamlJavadocBuilder.java:
```diff
@@ -0,0 +1,108 @@
+        this.processingEnv = processingEnv;
+    }
+
+    List<ImplData> getClassJavadocAsYamlOrNull(TypeElement classElement) {
+        Map<ElementKind, List<Element>> children = new HashMap<>();
```
Rename this - it should never return `null` but may return an empty list

gselzer replied:
I actually just got rid of the method.

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/YamlJavadocBuilder.java:
```diff
@@ -0,0 +1,108 @@
+        return implList;
+    }
+
+    private static List<ImplData> getJavadocsAsJson(List<Element> elements, Function<Element, ImplData> createDoc) {
```
Rename - we aren't returning JSON

gselzer replied:
Also rename the below method

gselzer replied:
Fixed

gselzer commented on scijava/scijava-javadoc-parser/src/main/java/org/scijava/javadoc/parser/YamlJavadocBuilder.java:
```diff
@@ -0,0 +1,108 @@
+
+    private final Function<Element, ImplData> fieldJavadocAsJson = (field) -> {
+        String javadoc = processingEnv.getElementUtils().getDocComment(field);
+        if (isBlank(javadoc)) {
+            return null;
+        }
```
This conditional would be covered by the next conditional - I'd delete it

gselzer replied:
Same with the method and class `Function`s

gselzer replied:
Not present anymore

gselzer commented on scijava/scijava-ops-engine/pom.xml:
```diff
@@ -186,6 +186,11 @@
 			<groupId>org.yaml</groupId>
 			<artifactId>snakeyaml</artifactId>
 		</dependency>
+		<!--TODO: Why is this necessary for the intellij build?-->
+		<dependency>
+			<groupId>org.slf4j</groupId>
+			<artifactId>slf4j-api</artifactId>
+		</dependency>
```
This can't actually be necessary, right??

gselzer replied:
It's not!

gselzer commented on scijava/scijava-ops-engine/src/main/java/module-info.java:
```diff
@@ -5,42 +5,25 @@
-	opens org.scijava.ops.engine.stats to therapi.runtime.javadoc;
-	opens org.scijava.ops.engine.util to therapi.runtime.javadoc;
-	opens org.scijava.ops.engine.math to therapi.runtime.javadoc;
+	opens org.scijava.ops.engine to org.scijava;
+	opens org.scijava.ops.engine.impl to org.scijava;
+	opens org.scijava.ops.engine.hint to org.scijava.javadoc.parser;
+	opens org.scijava.ops.engine.matcher.impl to org.scijava;
+	opens org.scijava.ops.engine.conversionLoss.impl to org.scijava;
```
Maybe file an issue to determine whether these `open`s are actually necessary.

Except for `org.scijava.ops.engine.hint` - why does that need to be opened to the Javadoc parser?

gselzer replied:
No opens anymore!

gselzer commented on scijava/scijava-ops-engine/src/main/java/module-info.java:
```diff
@@ -67,7 +49,7 @@
 
 	provides org.scijava.discovery.Discoverer with
- 		org.scijava.ops.engine.impl.TherapiOpInfoDiscoverer;
+			org.scijava.ops.engine.yaml.YAMLOpInfoDiscoverer;
```
Format

gselzer replied:
Formatted!

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/impl/OpFieldInfo.java:
```diff
@@ -81,7 +81,10 @@ public OpFieldInfo(final Object instance, final Field field, final String versio
 	}
 
-	public OpFieldInfo(final Object instance, final Field field, final String version, final Hints hints, final double priority, final String... names) {
+	public OpFieldInfo(final Object instance, final Field field,
+		final String version, final Hints hints, final double priority,
+		final String... names)
+	{
```
Was this formatted?

gselzer replied:
Unformatted!

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/struct/FieldOpDependencyMember.java:
```diff
@@ -57,7 +55,8 @@ public FieldOpDependencyMember(final Field field, final Class<?> structType) {
 	}
 
 	private static Producer<String> getDescription(Field f) {
-		return () -> RuntimeJavadoc.getJavadoc(f).getComment().toString();
+		// TODO
+		return () -> "";
```
Not sure why we'd need dependency descriptions - can this be deleted?

gselzer replied:
Yup!

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/struct/LazilyGeneratedFieldParameterData.java:
```diff
@@ -26,7 +22,7 @@
+ *
  * @author Gabriel Selzer
  */
 public class LazilyGeneratedFieldParameterData implements ParameterData {
```
There's the potential for even more simplification here - I'm assuming that the YAML `OpInfo` pathway doesn't call this anymore (it shouldn't because we should get the data directly from the YAML) - so maybe the other pathways could just inline the useful logic from this class, and then we could delete this?

gselzer replied:
Let's solve this as a part of scijava/scijava#138

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/struct/LazilyGeneratedMethodParameterData.java:
```diff
@@ -15,11 +15,7 @@
-import com.github.therapi.runtimejavadoc.RuntimeJavadoc;
 
 /**
  * Lazily generates the parameter data for a {@link List} of
```
The same thing I wrote for `LazilyGeneratedFieldParameterData` applies here

gselzer replied:
Let's solve this as a part of scijava/scijava#138

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/struct/MethodOpDependencyMemberParser.java:
```diff
@@ -30,7 +26,7 @@ public class MethodOpDependencyMemberParser implements
 {
 
-	private static final Map<Method, MethodJavadoc> methodDocMap = new ConcurrentHashMap<>();
+//	private static final Map<Method, MethodJavadoc> methodDocMap = new ConcurrentHashMap<>();
```
Delete it if we don't need it

gselzer replied:
Deleted

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/struct/MethodOpDependencyMemberParser.java:
```diff
@@ -85,18 +76,10 @@ private static void parseMethodOpDependencies(
-				return getDocData(annotatedMethod).getParams().get(j).getComment()
-					.toString();
-			};
+			// TODO
```
Do we need these `Producer`-style constructors anymore?

gselzer replied:
Nope!

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/yaml/AbstractYAMLOpInfoCreator.java:
```diff
@@ -16,24 +28,35 @@ public abstract class AbstractYAMLOpInfoCreator implements YAMLOpInfoCreator {
-        final String version = path.substring(path.indexOf('/') + 1);
+        // Parse source - start after the leading slash
+        final String srcString = identifier.getPath().substring(1);
+        // TODO: Parse version
+        final String version = "0.0";
```
This needs to be solved - we should probably find and cache the version from the POM, if we can't guarantee that the version can be included in the YAML file.

We may just want to create another issue, since the best solution is probably just to enable the SciJava Javadoc annotation processor to parse the version from the POM. I don't know if that means releasing scijava-meta, or duplicating code, meaning this may be a bigger decision that shouldn't hold up this PR. cc @ctrueden

gselzer replied:
I got versions figured out using a second argument to the annotation parser, which passes the version from the POM in the invocation of the annotation parser!

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/yaml/AbstractYAMLOpInfoCreator.java:
```diff
@@ -16,24 +28,35 @@ public abstract class AbstractYAMLOpInfoCreator implements YAMLOpInfoCreator {
+        final String srcString = identifier.getPath().substring(1);
+        // TODO: Parse version
+        final String version = "0.0";
+//        final String version = path.substring(path.indexOf('/') + 1);
```
Delete this if we aren't using it

gselzer replied:
Deleted

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/yaml/AbstractYAMLOpInfoCreator.java:
```diff
@@ -16,24 +28,35 @@ public abstract class AbstractYAMLOpInfoCreator implements YAMLOpInfoCreator {
         } else {
-            String namesString = (String) yaml.get("names");
-            names = namesString.split("\\s*,\\s*");
+            var tmp = tags.get("names");
```
Should also consider the case where there *is* no name - probably should throw an exception there, too.

gselzer replied:
We do now!

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/yaml/AbstractYAMLOpInfoCreator.java:
```diff
@@ -16,24 +28,35 @@ public abstract class AbstractYAMLOpInfoCreator implements YAMLOpInfoCreator {
+        final String version = "0.0";
+//        final String version = path.substring(path.indexOf('/') + 1);
+        final Map<String, Object> tags = subMap(yaml, "tags");
         // Parse names
```
Should partition this into its own method for readability

gselzer replied:
It is now!

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/yaml/AbstractYAMLOpInfoCreator.java:
```diff
@@ -16,24 +28,35 @@ public abstract class AbstractYAMLOpInfoCreator implements YAMLOpInfoCreator {
+        for (int i = 0; i < names.length; i++) {
+            names[i] = names[i].trim();
         }
         // Parse priority
```
This could also be its own method

gselzer replied:
It is now!

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/yaml/AbstractYAMLOpInfoCreator.java:
```diff
@@ -42,13 +65,108 @@ else if (p instanceof String) {
         } catch (Exception e) {
             throw new RuntimeException(e);
         }
+        if (tags.containsKey("parameters")) {
+            List<Map<String, Object>> params =
+                (List<Map<String, Object>>) tags.get("parameters");
+            Iterator<Map<String, Object>> paramItr = params.iterator();
+
+            List<Member<?>> members = info.struct().members();
+            for (int i = 0; i < members.size(); i++) {
+                Member<?> m = members.get(i);
+                if (m.isInput() || m.isOutput()) {
+                    if (!paramItr.hasNext()) break;
+                    Map<String, Object> paramMap = paramItr.next();
+                    members.set(i, wrapMember.apply(m, paramMap));
+                }
+            }
+        }
+
+        return info;
```
This feels like it could be improved - do we really have to wrap members? Can we not just create them correctly the first time?

gselzer replied:
Ideally, we'd find a way to pass through parameter data in the `OpClassInfo` etc. constructor - but I think it's a change beyond the scope of this PR - I'll file a new PR.

gselzer replied:
I wrote scijava/scijava#138 to address this.

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/yaml/YAMLOpInfoDiscoverer.java:
```diff
@@ -92,38 +94,11 @@ private void parse(List<OpInfo> infos, final URL url)
-				System.out.println("Could not add op " + identifier + ":");
-				e.printStackTrace();
+				throw new IllegalArgumentException(e);
+//				System.out.println("Could not add op " + identifier + ":");
+//				e.printStackTrace();
```
Delete

gselzer replied:
Removed

gselzer commented on scijava/scijava-ops-engine/src/main/resources/META-INF/services/org.scijava.discovery.Discoverer:
Should this be deleted?

gselzer replied:
Nope - just updated it

gselzer commented on scijava/scijava-ops-engine/src/test/java/org/scijava/ops/engine/yaml/ops/YAMLClassOp.java:
```diff
@@ -7,6 +7,10 @@
 
 /**
  * An example Op, registered by YAML into the {@link OpEnvironment}
+ * @implNote op names=example.add
+ * @input aDouble the first input
+ * @input bDouble the second input
+ * @output the result
```
Man, putting the implementation details here feels like a vulnerability to action-at-a-distance errors - compilers and/or IDE tools won't see the correlation between this and (a) the Javadoc on the functional method and (b) the functional method parameters.

But, at the same time, keeping  the YAML parser general means probably not special-casing logic to pull tags from functional methods.

If you write an Op as a class, you probably don't want to copy your javadoc up to the class, right?

I don't know, does anyone else have opinions? @ctrueden @hinerm

ctrueden replied:
Solution: discontinue support for class-based op implementations! :clown_face:

ctrueden replied:
In seriousness: I assume there are reasons why class-based op implementations are sometimes necessary/preferred over method-based ones? (I know for field-based ones, we cannot support e.g. op dependencies.)

ctrueden replied:
My half-baked opinion is: how about changing class-based op implementations to read that stuff from the functional method, and *not* from the class? The only weirdness there is that people will need to put javadoc on an `@Override`n method, but that's certainly a permitted thing to do, and would then be documenting the actual function.

gselzer replied:
> In seriousness: I assume there are reasons why class-based op implementations are sometimes necessary/preferred over method-based ones? (I know for field-based ones, we cannot support e.g. op dependencies.)

I don't think there's any reason why it's necessary, but the preference thing I think is important. I do not want to make [`RichardsonLucyC`](https://github.com/scijava/incubator/blob/105acb4fd5aaccac673ef7ed23f9b6b5acba17e9/imagej/imagej-ops2/src/main/java/net/imagej/ops2/deconvolve/RichardsonLucyC.java#L62) a `Method` - imagine how unwieldy that'd be as a single method :exploding_head:

gselzer replied:
> My half-baked opinion is: how about changing class-based op implementations to read that stuff from the functional method, and _not_ from the class? The only weirdness there is that people will need to put javadoc on an `@Override`n method, but that's certainly a permitted thing to do, and would then be documenting the actual function.

Well, this is what was done with therapi, and it is imo the nicest way to do things. And we could make a specialization on FunctionalInterfaces that pulls javadoc tags from *both* the class and the functional method - the only thing to decide then is where the logic lives.

Right now, there's probably sufficient logic within SciJava Types for that, so the question is whether SciJava Javadoc Parser should depend on SciJava Types. That'd make the build even more of a nightmare, so I'm tempted to just fork the necessary logic into this module...

ctrueden replied:
I don't understand why the scijava javadoc parser needs to reason about functional interface methods? Doesn't it just index everything declared in the javadoc?

gselzer replied:
The SciJava Javadoc Parser is more specialized, only looking for and storing elements with `@implNote` implementations. In the case of `Class` elements, the `@implNote` will be on the class. Thus, to pull any tags off of the functional method, we'll have to:

1) Determine the functional interface implemented by the Class
2) Determine the functional method implemented
3) Find that method and grab all tags.

The first two steps can be done through SciJava Types.

ctrueden replied:
In order to support method-based op declarations, the SciJava javadoc parser is already responding to implNote declarations on methods, right? Can we not simply use the already-recorded implNote details on the method, without doing all the work you describe? In other words: the burden of understanding how to interpret the implNote and related indexed metadata would remain within the Ops framework, not the javadoc parser? So then it's just a matter of disambiguating the two different kinds of methods that will have implNotes: either an op-method kind, or a functional-interface-method-of-op-class kind?

ctrueden replied:
Another way to articulate my more general concern here: the javadoc parser is writing out an Ops-style YAML file directly, right? If so, it's not really the "scijava javadoc parser" at that point, but rather the *Ops* javadoc parser. No? Or... does the scijava javadoc parser write out some other intermediate thing? How exactly does it work?

gselzer replied:
> So then it's just a matter of disambiguating the two different kinds of methods that will have implNotes: either an op-method kind, or a functional-interface-method-of-op-class kind?

Hmm, yeah, that could work. Maybe I'll give it a shot.

> Another way to articulate my concern here: the javadoc parser is writing out an Ops-style YAML file directly, right? If so, it's not really the "scijava javadoc parser" at that point, but rather the Ops javadoc parser. No? Or... does the scijava javadoc parser write out some other intermediate thing? How exactly does it work?

It's more general than ops, for each `@implNote blah`, a file `blah.yaml` is output in the target folder containing all elements (classes, fields, methods, may also want to consider Java 16 records) with that tag. I explain it pretty succinctly and with an explanation in the PR description up top.

gselzer replied:
Although, now that I look at it, I think the example yaml I put in the PR description is a little different than what is recorded now...

gselzer replied:
One other thing I'll note: I am really interested in others' opinions on whether this annotation processor *would* be useful in contexts beyond ops - if you think it is not, then yeah, we should probably rename it and specialize a bit for Ops.

gselzer replied:
We no longer need to put the tags on the class!

gselzer reviewed (commented):
Phew! What a PR!

gselzer committed:
Restore IterableInterval copier

Without it, it's really difficult to match the request in the tests.
Since it isn't hurting anything to have it, we might as well keep it.

gselzer committed:
Convert command line arg to boolean-like

Since we're storing far less information now, I think it's a lot easier
to just have an opt-in toggle, and I don't see anyone wanting to only
parse some packages.

gselzer commented:
Some additional thoughts from discussing with @hinerm and @ctrueden:

* Let's rename it the SciJava Op Indexer, and specialize it on Ops. We can always copy the code to make something broader than that.
* Let's try to make the `type` tag *optional* for methods - if the user leaves it out, then we can map to it the appropriate `Computer` or `Function`.
* Let's move a lot of the parameters out of the `tags` map, and leave it for only truly optional stuff (like the OpMethod type) - authors, name, parameters, etc. are all required for all Ops, and it's fine since we are specializing on Ops.

gselzer committed:
Rename to SciJava Ops Indexer

@ctrueden and @hinerm and I decided that we should specialize this
module on Ops, because we don't yet have a use case beyond Ops.

hinerm commented:
@gselzer with https://github.com/scijava/incubator/pull/86 merged the docs will have to be updated as well. I think they can be as part of this PR, or split out to a new issue/branch if you prefer

gselzer committed:
"Op-ify" the yaml

This makes the logic more complex - let's clean it up in the following
commits

gselzer committed:
Rename annotation processor option

gselzer committed:
Move inner class HACK out of SciJava Common

gselzer committed:
Rename and add documentation to OpImplData

gselzer committed:
Cleanup JavadocAnnotationProcessor

gselzer committed:
Clean up OpImplData

A lot (i.e. all) of the fields are shared, so let's move them upstream

gselzer committed:
Enable YAML Ops to infer their type

gselzer committed:
Require correct no. params for implNote ops

gselzer committed:
Validate Ops declared in javadoc

gselzer committed:
Add Op version to the parser

gselzer committed:
Fix indentation

gselzer committed:
Rename Annotation Processor

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/struct/LazilyGeneratedMethodParameterData.java:
```diff
@@ -15,11 +15,7 @@
 import org.scijava.ops.engine.reduce.ReductionUtils;
 import org.scijava.ops.spi.OpDependency;
 import org.scijava.struct.FunctionalMethodType;
-import org.scijava.struct.ItemIO;
```
Do we even need this class anymore?

gselzer replied:
See scijava/scijava#131

gselzer committed:
Add comment explaining the false return

gselzer committed:
Combine OpImplNoteParser and YamlJavadocBuilder

gselzer committed:
Remove commented out log message

gselzer committed:
Update Services

gselzer committed:
Add LICENSE for adapted Therapi code

gselzer committed:
Remove InvalidOpJavadocException

Wasn't being used

gselzer committed:
Various small clean-up changes

gselzer committed:
Format module-info.java

gselzer committed:
Clean up OpDependencyMemberParsers

gselzer committed:
Remove TODO in YAML version parsing

gselzer committed:
Throw IAE when Op yaml has no name

gselzer committed:
Partition name parser into its own method

gselzer commented on imagej/imagej-ops2/src/test/java/net/imagej/ops2/copy/CopyTypeTest.java:
```diff
@@ -30,6 +30,8 @@
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
+import java.util.regex.Pattern;
```
Remove this

gselzer replied:
Removed!

gselzer commented on pom.xml:
```diff
@@ -152,6 +150,11 @@
 					</dependency>
 				</dependencies>
 			</plugin>
+			<plugin>
+				<groupId>com.alexecollins.maven.plugin</groupId>
+				<artifactId>script-maven-plugin</artifactId>
+				<version>${script-maven-plugin.version}</version>
+			</plugin>
```
Why does the diff mark this as new, and why does the script this runs show as formatted strangely?

gselzer replied:
After a second look, I think it's all fine - without this, I don't think the script runs.

gselzer commented on scijava/scijava-ops-engine/src/main/java/module-info.java:
```diff
@@ -5,42 +5,19 @@
 	requires org.scijava.function;
 	requires org.scijava.log2;
 	requires org.scijava.meta;
+	requires org.scijava.parse2;
```
Do we need parse2 for something here, or can we remove it (and from `module-info.vm`)?

gselzer replied:
Nope, we don't need it!

gselzer commented on scijava/scijava-ops-indexer/src/main/java/org/scijava/ops/indexer/OpFieldImplData.java:
```diff
@@ -0,0 +1,101 @@
+	 * @param additionalTags the tags pertaining exclusively to {@link Field}s
+	 */
+	@Override
+	void parseAdditionalTags(Element source, List<String[]> additionalTags) {
```
A little documentation throughout this method would be awesome

gselzer replied:
Cleaned up in 53e6e345

gselzer commented on scijava/scijava-ops-indexer/src/main/java/org/scijava/ops/indexer/OpImplData.java:
```diff
@@ -0,0 +1,167 @@
+		return remainingTags;
+	}
+
+	abstract void parseAdditionalTags(Element source, List<String[]> additionalTags);
```
Might want a quick Javadoc explaining what this abstract method is for.

gselzer replied:
Added in f230f8d8

gselzer commented on scijava/scijava-ops-indexer/src/main/java/org/scijava/ops/indexer/OpImplNoteParser.java:
```diff
@@ -0,0 +1,200 @@
+		return Collections.singleton(PARSE_OPS);
+	}
+
+	private final Function<Element, Optional<OpImplData>> elementToImplData = (
```
Might as well make this a function

gselzer replied:
Converted in b49fe804

gselzer commented on scijava/scijava-ops-indexer/src/main/java/org/scijava/ops/indexer/OpMethodImplData.java:
```diff
@@ -0,0 +1,150 @@
+			}
+		}
+
+		for (int i = 0; i < functionalParams.size(); i++) {
```
I think this could be an enhanced for loop?

gselzer replied:
Can't use an enhanced for loop, but we can just remove the for loop entirely! Refactored in 7b53b90d

gselzer commented on scijava/scijava-ops-indexer/src/main/java/org/scijava/ops/indexer/ProcessingUtils.java:
```diff
@@ -0,0 +1,112 @@
+		throw new AssertionError("not instantiable");
+	}
+
+	public static void printProcessingException(Element source, Throwable t,
```
Doc here would be nice

gselzer replied:
Added in 2de12a13

gselzer commented on scijava/scijava-ops-indexer/src/main/java/org/scijava/ops/indexer/ProcessingUtils.java:
```diff
@@ -0,0 +1,112 @@
+			"Exception parsing source + " + source + ": " + sw);
+	}
+
+	public static ExecutableElement findFunctionalMethod(
```
Doc here would be nice

gselzer replied:
Added in 2de12a13

gselzer reviewed (commented):
Super close to done here!

gselzer committed:
Remove hints from method

gselzer committed:
Remove SciJava Parse2 dependency

It wasn't used

gselzer committed:
OpFieldImplData: Document parseAdditionalTags

gselzer committed:
OpImplData: Javadoc for parsing additional tags

gselzer committed:
OpImplNoteParser: Convert Field to Method

No real reason to keep it as a field

gselzer committed:
OpMethodImplData: simplify

gselzer committed:
ProcessingUtils: Add documentation

gselzer committed:
Write README for the Ops Indexer

gselzer committed:
Improve ImplNote parser regex
